scripts/entity_linking.py

```diff
@@ -17,93 +17,150 @@
-from REL.REL.entity_disambiguation import EntityDisambiguation
-from REL.REL.ner import NERBase, Span
+import sys
+from REL.mention_detection import MentionDetectionBase
+from REL.utils import process_results, split_in_words
+from REL.entity_disambiguation import EntityDisambiguation
+from REL.ner import Span
 from wikimapper import WikiMapper
-
+from typing import Dict, List, Union, Tuple
+from tqdm import tqdm
 
 # Spacy Mention Detection class which overrides the NERBase class in the REL entity linking process
-class NERSpacy(NERBase):
-    def __init__(self):
-        # we only want to link entities of specific types
-        self.ner_labels = ['PERSON', 'NORP', 'FAC', 'ORG', 'GPE', 'LOC', 'PRODUCT', 'EVENT', 'WORK_OF_ART',
-                           'LAW', 'LANGUAGE', 'DATE', 'TIME', 'MONEY', 'QUANTITY']
-
-    # mandatory function which overrides NERBase.predict()
-    def predict(self, doc):
-        mentions = []
-        for ent in doc.ents:
-            if ent.label_ in self.ner_labels:
-                mentions.append(Span(ent.text, ent.start_char, ent.end_char, 0, ent.label_))
-        return mentions
+class NERSpacyMD(MentionDetectionBase):
+	def __init__(self, base_url:str, wiki_version:str, spacy_model:str):
+		super().__init__(base_url, wiki_version)
+		# we only want to link entities of specific types
+		self.ner_labels = ['PERSON', 'NORP', 'FAC', 'ORG', 'GPE', 'LOC', 'PRODUCT', 'EVENT', 'WORK_OF_ART',
+						   'LAW', 'LANGUAGE', 'DATE', 'TIME', 'MONEY', 'QUANTITY']
+		self.spacy_model = spacy_model
+		spacy.prefer_gpu()
+		self.tagger = spacy.load(spacy_model)
+
+	# mandatory function which overrides NERBase.predict()
+	def predict(self, doc: spacy.tokens.Doc) -> List[Span]:
+		spans = []
+		for ent in doc.ents:
+			if ent.label_ in self.ner_labels:
+				spans.append(Span(ent.text, ent.start_char, ent.end_char, 0, ent.label_))
+		return spans
+
+	"""
+	Responsible for finding mentions given a set of documents in a batch-wise manner. More specifically,
+	it returns the mention, its left/right context and a set of candidates.
+	:return: Dictionary with mentions per document.
+	"""
+
+	def find_mentions(self, dataset: Dict[str, str]) -> Tuple[Dict[str, List[Dict]], int]:
+		results = {}
+		total_ment = 0
+		for i, doc in tqdm(enumerate(dataset), desc='Finding mentions', total=len(dataset)):
+			result_doc = []
+			doc_text = dataset[doc]
+			spacy_doc = self.tagger(doc_text)
+			spans = self.predict(spacy_doc)
+			for entity in spans:
+				text, start_pos, end_pos, conf, tag = (
+					entity.text,
+					entity.start_pos,
+					entity.end_pos,
+					entity.score,
+					entity.tag,
+				)
+				m = self.preprocess_mention(text)
+				cands = self.get_candidates(m)
+				if len(cands) == 0:
+					continue
+				total_ment += 1
+				# Re-create ngram as 'text' is at times changed by Flair (e.g. double spaces are removed).
+				ngram = doc_text[start_pos:end_pos]
+				left_ctxt = split_in_words(doc_text[:start_pos])
```

Comment:
Maybe you can merge these three lines:

left_ctxt = " ".join(split_in_words(doc_text[:start_pos])[-100:])
same for right_ctxt